volume/volumefs/restore.go

```diff
@@ -0,0 +1,297 @@
+package volumefs
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"sync"
+	"time"
+
+	"github.com/kopia/kopia/fs"
+	"github.com/kopia/kopia/volume"
+
+	"github.com/pkg/errors"
+)
+
+// Restore constants.
+const (
+	DefaultRestoreConcurrency = 4
+)
+
+// RestoreStats contains restore statistics
+type RestoreStats struct {
+	BytesWritten  int64
+	NumDirs       int
+	NumFiles      int
+	NumDirEntries int64
+	MinBlockAddr  int64
+	MaxBlockAddr  int64
+}
+
+func (rs *RestoreStats) init(f *Filesystem) {
+	rs.MinBlockAddr = f.maxBlocks + 1
+	rs.MaxBlockAddr = -1
+}
+
+// Restore extracts a snapshot to a volume.
+// The provided volume manager must provide a BlockWriter interface.
+func (f *Filesystem) Restore(ctx context.Context, prevSnapID string) (RestoreStats, error) {
```

Comment:
Provide an option to simply traverse the snapshot and collect traversal statistics